src/platform/graphics/blend-state.js

```diff
@@ -0,0 +1,222 @@
+     * - {@link BLENDMODE_ONE_MINUS_SRC_COLOR}
+     * - {@link BLENDMODE_DST_COLOR}
+     * - {@link BLENDMODE_ONE_MINUS_DST_COLOR}
+     * - {@link BLENDMODE_SRC_ALPHA}
+     * - {@link BLENDMODE_SRC_ALPHA_SATURATE}
+     * - {@link BLENDMODE_ONE_MINUS_SRC_ALPHA}
+     * - {@link BLENDMODE_DST_ALPHA}
+     * - {@link BLENDMODE_ONE_MINUS_DST_ALPHA}
+     * - {@link BLENDMODE_CONSTANT}
+     * - {@link BLENDMODE_ONE_MINUS_CONSTANT}
+     *
+     * All op parameters can take the following values:
+     *
+     * - {@link BLENDEQUATION_ADD}
+     * - {@link BLENDEQUATION_SUBTRACT}
+     * - {@link BLENDEQUATION_REVERSE_SUBTRACT}
+     * - {@link BLENDEQUATION_MIN}
+     * - {@link BLENDEQUATION_MAX}
+     *
+     * Note that MIN and MAX operations on WebGL platform require either EXT_blend_minmax or WebGL2
+     * to work (check device.extBlendMinmax).
+     *
+     * @param {boolean} blend - Enables or disables blending. Defaults to false.
+     * @param {number} colorOp - Configures color blending operation. Defaults to
+     * {@link BLENDEQUATION_ADD}.
+     * @param {number} colorSrcFactor - Configures source color blending factor. Defaults to
+     * {@link BLENDMODE_ONE}.
+     * @param {number} colorDstFactor - Configures destination color blending factor. Defaults to
+     * {@link BLENDMODE_ZERO}.
+     * @param {number} alphaOp - Configures alpha blending operation. Defaults to
+     * {@link BLENDEQUATION_ADD}.
+     * @param {number} alphaSrcFactor - Configures source alpha blending factor. Defaults to
+     * {@link BLENDMODE_ONE}.
+     * @param {number} alphaDstFactor - Configures destination alpha blending factor. Defaults to
+     * {@link BLENDMODE_ZERO}.
+     * @param {boolean} redWrite - True to enable writing of the red channel and false otherwise.
+     * Defaults to true.
+     * @param {boolean} greenWrite - True to enable writing of the green channel and false
+     * otherwise. Defaults to true.
+     * @param {boolean} blueWrite - True to enable writing of the blue channel and false otherwise.
+     * Defaults to true.
+     * @param {boolean} alphaWrite - True to enable writing of the alpha channel and false
+     * otherwise. Defaults to true.
+     */
+    constructor(blend = false, colorOp = BLENDEQUATION_ADD, colorSrcFactor = BLENDMODE_ONE, colorDstFactor = BLENDMODE_ZERO,
+        alphaOp, alphaSrcFactor, alphaDstFactor,
+        redWrite = true, greenWrite = true, blueWrite = true, alphaWrite = true) {
+        this.setColorBlend(colorOp, colorSrcFactor, colorDstFactor);
+        this.setAlphaBlend(alphaOp ?? colorOp, alphaSrcFactor ?? colorSrcFactor, alphaDstFactor ?? colorDstFactor);
+        this.setColorWrite(redWrite, greenWrite, blueWrite, alphaWrite);
+        this.blend = blend;
+    }
+
+    /**
+     * Enables or disables blending.
+     *
+     * @type {boolean}
+     */
+    set blend(value) {
+        this.target0 = BitPacking.set(this.target0, value ? 1 : 0, blendShift);
```

Comment:
Isn't this doing more than necessary?

1) The bit operations involve the property lookup in the `BitPacking` object, which could be just exported as normal functions, e.g. named `bitPackingSet`

2) I am not sure about the exact working of `BitPacking.set`, but isn't this line the same as:

```js
this.target0 |= (value ? 1 : 0) << blendShift;
```
?

I did some testing and it seems similiar:

```js
var diffResults = 0;
for (var start = -100000; start < 100000; start++) {
    for (var i=-100; i<100; i++) {
        const a = BitPacking.set(start, 1, i);
        const b = start | 1 << i;
        if (a != b) {
            diffResults++;
        }
    }
}
console.log("diffResults", diffResults); // 0
```

Reply:
1. I did some performance testing, and having those functions inside an object was acceptable: https://github.com/playcanvas/engine/pull/5071
2. it's not the same. BitPacking.set allows values to be set to 0 or 1, but your implementation with `\=' only allows it to be set to 1, but never cleared to 0.

Reply:
Thank you, I was blind to the 0 case :eyes: 

```js
var diffResults = 0;
var sameResults = 0;
for (var start = -100000; start < 100000; start++) {
    for (var i=-100; i<100; i++) {
        const a = BitPacking.set(start, 0, i);
        const b = start | 0 << i;
        if (a != b) {
            diffResults++;
        } else {
            sameResults++;
        }
    }
}
console.log("diffResults", diffResults, "sameResults", sameResults);
// diffResults 20000000 sameResults 20000000
```